File: dotnet/Falcon.PluginAPI/Entities/Player.cs
using System;
using Falcon.Interop;

namespace Falcon
{
    public sealed unsafe class Player : IEquatable<Player>
    {
        private readonly FalconPlayer* _handle;

        internal Player(FalconPlayer* handle)
        {
            _handle = handle;
        }

        public nint Handle
        {
            get
            {
                return (nint)_handle;
            }
        }

        public bool IsValid
        {
            get
            {
                return _handle != null;
            }
        }

        public string Name
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->playerName(_handle));
            }
        }

        public bool IsOperator
        {
            get
            {
                return NativeApi.Table->playerIsOperator(_handle) != 0;
            }
            set
            {
                NativeApi.Table->playerSetOperator(_handle, value ? 1 : 0);
            }
        }

        public Entity Entity
        {
            get
            {
                return new Entity(NativeApi.Table->playerEntity(_handle));
            }
        }

        public GameMode GameMode
        {
            get
            {
                return (GameMode)NativeApi.Table->playerGameMode(_handle);
            }
            set
            {
                NativeApi.Table->playerSetGameMode(_handle, (uint)value);
            }
        }

        public string Xuid
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->playerXuid(_handle));
            }
        }

        public string Uuid
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->playerUuid(_handle));
            }
        }

        public string Address
        {
            get
            {
                return NativeApi.Text(NativeApi.Table->playerAddress(_handle));
            }
        }

        public float Food
        {
            get
            {
                return NativeApi.Table->playerFood(_handle);
            }
            set
            {
                NativeApi.Table->playerSetFood(_handle, value);
            }
        }

        public int XpLevel
        {
            get
            {
                return NativeApi.Table->playerXpLevel(_handle);
            }
            set
            {
                NativeApi.Table->playerSetXpLevel(_handle, value);
            }
        }

        public PlayerInventory Inventory
        {
            get
            {
                return new PlayerInventory(_handle);
            }
        }

        internal FalconPlayer* Raw
        {
            get
            {
                return _handle;
            }
        }

        public static bool operator ==(Player? left, Player? right)
        {
            if (left is null)
            {
                return right is null;
            }

            return left.Equals(right);
        }

        public static bool operator !=(Player? left, Player? right)
        {
            return !(left == right);
        }

        public void SendMessage(string message)
        {
            fixed (byte* text = NativeApi.Utf8(message))
            {
                NativeApi.Table->playerSendMessage(_handle, text);
            }
        }

        public void Kick(string reason)
        {
            fixed (byte* text = NativeApi.Utf8(reason))
            {
                NativeApi.Table->playerKick(_handle, text);
            }
        }

        public void SendTitle(string title, string subtitle = "")
        {
            fixed (byte* titleText = NativeApi.Utf8(title))
            fixed (byte* subtitleText = NativeApi.Utf8(subtitle))
            {
                NativeApi.Table->playerSendTitle(_handle, titleText, subtitleText);
            }
        }

        public void SendActionBar(string message)
        {
            fixed (byte* text = NativeApi.Utf8(message))
            {
                NativeApi.Table->playerSendActionBar(_handle, text);
            }
        }

        public bool HasPermission(string node)
        {
            fixed (byte* text = NativeApi.Utf8(node))
            {
                return NativeApi.Table->playerHasPermission(_handle, text) != 0;
            }
        }

        public void SetPermission(string node, bool value = true)
        {
            FalconPlugin* plugin = PluginScope.Require();
            fixed (byte* text = NativeApi.Utf8(node))
            {
                NativeApi.Table->playerSetPermission(plugin, _handle, text, value ? 1 : 0);
            }
        }

        public void UnsetPermission(string node)
        {
            FalconPlugin* plugin = PluginScope.Require();
            fixed (byte* text = NativeApi.Utf8(node))
            {
                NativeApi.Table->playerUnsetPermission(plugin, _handle, text);
            }
        }

        public bool SendPacket(uint packetId, ReadOnlySpan<byte> data)
        {
            fixed (byte* bytes = data)
            {
                return NativeApi.Table->playerSendPacket(_handle, packetId, bytes, (uint)data.Length) != 0;
            }
        }

        public bool Equals(Player? other)
        {
            return other is not null && other._handle == _handle;
        }

        public override bool Equals(object? other)
        {
            return Equals(other as Player);
        }

        public override int GetHashCode()
        {
            return ((nint)_handle).GetHashCode();
        }

        internal static Player? Wrap(FalconPlayer* handle)
        {
            if (handle == null)
            {
                return null;
            }

            return new Player(handle);
        }
    }
}

File: java/src/main/java/falcon/api/Player.java
package falcon.api;

import falcon.api.internal.Interop;
import falcon.api.internal.NativeApi;
import falcon.api.internal.PluginContext;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;

public final class Player {
    private final PluginContext mContext;
    private final MemorySegment mHandle;

    public Player(PluginContext context, MemorySegment handle) {
        mContext = context;
        mHandle = handle == null ? MemorySegment.NULL : handle;
    }

    public boolean valid() {
        return !Interop.isNull(mHandle);
    }

    public String name() {
        return Interop.string(api().playerName(mHandle));
    }

    public boolean isOperator() {
        return api().playerIsOperator(mHandle) != 0;
    }

    public void setOperator(boolean value) {
        api().playerSetOperator(mHandle, value ? 1 : 0);
    }

    public void sendMessage(String message) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerSendMessage(mHandle, Interop.text(arena, message));
        }
    }

    public void kick(String reason) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerKick(mHandle, Interop.text(arena, reason));
        }
    }

    public Entity entity() {
        return new Entity(mContext, api().playerEntity(mHandle));
    }

    public GameMode gameMode() {
        return GameMode.of(api().playerGameMode(mHandle));
    }

    public void setGameMode(GameMode gameMode) {
        api().playerSetGameMode(mHandle, gameMode.value());
    }

    public String xuid() {
        return Interop.string(api().playerXuid(mHandle));
    }

    public String uuid() {
        return Interop.string(api().playerUuid(mHandle));
    }

    public String address() {
        return Interop.string(api().playerAddress(mHandle));
    }

    public void sendTitle(String title) {
        sendTitle(title, "");
    }

    public void sendTitle(String title, String subtitle) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerSendTitle(mHandle, Interop.text(arena, title), Interop.text(arena, subtitle));
        }
    }

    public void sendActionBar(String message) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerSendActionBar(mHandle, Interop.text(arena, message));
        }
    }

    public float food() {
        return api().playerFood(mHandle);
    }

    public void setFood(float food) {
        api().playerSetFood(mHandle, food);
    }

    public int xpLevel() {
        return api().playerXpLevel(mHandle);
    }

    public void setXpLevel(int level) {
        api().playerSetXpLevel(mHandle, level);
    }

    public PlayerInventory inventory() {
        return new PlayerInventory(mHandle);
    }

    public boolean hasPermission(String node) {
        try (Arena arena = Arena.ofConfined()) {
            return api().playerHasPermission(mHandle, Interop.text(arena, node)) != 0;
        }
    }

    public void setPermission(String node) {
        setPermission(node, true);
    }

    public void setPermission(String node, boolean value) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerSetPermission(mContext.handle(), mHandle, Interop.text(arena, node), value ? 1 : 0);
        }
    }

    public void unsetPermission(String node) {
        try (Arena arena = Arena.ofConfined()) {
            api().playerUnsetPermission(mContext.handle(), mHandle, Interop.text(arena, node));
        }
    }

    public boolean sendPacket(int packetId, byte[] data) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment bytes = MemorySegment.NULL;
            if (data.length > 0) {
                bytes = arena.allocateFrom(ValueLayout.JAVA_BYTE, data);
            }
            return api().playerSendPacket(mHandle, packetId, bytes, data.length) != 0;
        }
    }

    public MemorySegment handle() {
        return mHandle;
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof Player player && player.mHandle.address() == mHandle.address();
    }

    @Override
    public int hashCode() {
        return Long.hashCode(mHandle.address());
    }

    private static NativeApi api() {
        return Interop.api();
    }
}
